### PR TITLE
Bring your own Data plugins for flytekit

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
Flytekit depends on some additional cloud-provided tooling for its typical operation.  Specifically, when downloading or uploading task inputs or outputs, or when downloading or uploading blob-storage-based literals (files, schemas, folders), it makes use cloud provider CLI's like `aws cli` and `gsutil` to handle the transfer.  This is a bit of a legacy decision, which was made prior to python3 (in python2 days). Without asyncio, the performance of the downloader would not match the ones offered by the CLIs. 
Moreover, using the CLI's actually removes coupling with cloud provider SDKs and hence keep the flytekit core dependency free.  This has really worked well and we will continue to keep it going forward. But, we would love for contributors to bring in new ideas on handling this data - for example use non blob stores (like bigtable) for the metadata, or write native clients for managing the data or use a mature project like fsspec

The PR introduces a new `Persistence` Interface and provides `extras` for `aws cli` and `gsutil`. Users have to simply install
`pip install awscli` or `pip install gsutil` to use these extras. Moreover, the Persistence layer allows plugins and overriding the default data extras that are bundled.

This [RFC](https://docs.google.com/document/d/1nmARAWkaPq5VrSBULU0z7Fyihmq6aAeicnpUWvn1p0E/edit#heading=h.wy5gwufle4ch) provides the design intuition. An example data plugin will be added as a follow on, and we will use automatic plugin discovery to autoload the plugin.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
* Added a `DataPersistence` base class. This encapsulates the behavior of a blob storage provider and has things like get/list/put, etc.
* Added a `DataPersistencePlugins` class - this is the keeper of plugins, similar to how the `TypeEngine` keeps track of type transformers
* Added a `DiskPersistence` that inherits the `DataPersistence` class.  The local FS can be thought of as a storage provider - this object fakes that.  Paths that begin with `/` or `file://` by default will use this persistence plugin.
* Moved `FileAccessProvider` to under the `core/` folder.  This object is the high level interface that the rest of the flytekit platform and its plugins interact with.  Users/contributors should not need to work with the underlying `DataPersistence` classes (unless of course you're implementing one).

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/809

### CODE REVIEW

File: flytekit/core/data_persistence.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+======================================
+
+.. currentmodule:: flytekit.core.data_persistence
+
+The Data persistence module is used by core flytekit and most of the core TypeTransformers to manage data fetch & store,
+between the durable backend store and the runtime environment. This is designed to be a pluggable system, with a default
+simple implementation that ships with the core.
+
+.. autosummary::
+   :toctree: generated/
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:13]:
```suggestion
.. autosummary::
   :toctree: generated/
   :template: custom.rst
   :nosignatures:
```

File: flytekit/core/data_persistence.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        Constructs a path in the format <base><delim>*args
+        delim is dependent on the storage medium.
+        each of the args is joined with the delim
+        """
+        pass
+
+
+class DataPersistencePlugins(object):
+    """
+    This is core plugin engine that stores all DataPersistence plugins. To add a new plugin use
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:105]:
```suggestion
    This is the core plugin engine that stores all DataPersistence plugins. To add a new plugin use
```

File: flytekit/core/data_persistence.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    This exception is raised for all methods when a method is not supported by the data persistence layer
+    """
+
+    def __init__(self, message: str):
+        super(UnsupportedPersistenceOp, self).__init__(message)
+
+
+class DataPersistence(object):
+    """
+    Base abstract type for all DataPersistence operations. This can be plugged in using the flytekitplugins architecture
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:50]:
```suggestion
    Base abstract type for all DataPersistence operations. This can be extended using the flytekitplugins architecture
```

File: flytekit/core/data_persistence.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    def exists(self, path: str) -> bool:
+        """
+        Returns true if the given path exists, else false
+        """
+        pass
+
+    @abstractmethod
+    def get(self, from_path: str, to_path: str, recursive: bool = False):
+        """
+        Retrieves a data from from_path and writes to the given to_path (to_path is locally accessible)
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:81]:
```suggestion
        Retrieves data from from_path and writes to the given to_path (to_path is locally accessible)
```

File: flytekit/core/data_persistence.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+    These plugins should always be registered. Follow the plugin registration guidelines to auto-discover your plugins.
+    """
+
+    _PLUGINS: Dict[str, typing.Type[DataPersistence]] = {}
+
+    @classmethod
+    def register_plugin(cls, protocol: str, plugin: typing.Type[DataPersistence], force: bool = False):
+        """
+        Registers the supplied plugin for the specified protocol if one does not already exists.
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:119]:
```suggestion
        Registers the supplied plugin for the specified protocol if one does not already exist.
```

File: flytekit/core/data_persistence.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    def is_supported_protocol(cls, protocol: str) -> bool:
+        """
+        Returns true if the given protocol is has a registered plugin for it
+        """
+        return protocol in cls._PLUGINS
+
+
+class DiskPersistence(DataPersistence):
+    """
+    The simplest form of persistence that is available with default flytekit - Disk based persistence.
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:164]:
```suggestion
    The simplest form of persistence that is available with default flytekit - Disk-based persistence.
```

File: flytekit/core/data_persistence.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+class FileAccessProvider(object):
+    """
+    This is the class that is available through the FlyteContext and can be used for persisting data to the remote
+    durable store.
+    """
+
+    def __init__(self, local_sandbox_dir: Union[str, os.PathLike], raw_output_prefix: str):
+        # Local access
+        if local_sandbox_dir is None or local_sandbox_dir == "":
+            raise Exception("Can't use empty path")
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:240]: a more specific error?

```suggestion
            raise TypeError("Can't use empty path")
```

File: flytekit/core/data_persistence.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        self._raw_output_prefix = raw_output_prefix
+
+    @staticmethod
+    def is_remote(path: Union[str, os.PathLike]) -> bool:
+        """
+        Deprecated. Lets find a replacement
+        """
+        if path.startswith("/") or path.startswith("file://"):
+            return False
+        return True
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:256]: nit - conciseness: feel free to ignore

```suggestion
        return not (path.startswith("/") or path.startswith("file://"))
```

File: flytekit/core/data_persistence.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    @property
+    def local_sandbox_dir(self) -> os.PathLike:
+        return self._local_sandbox_dir
+
+    @property
+    def local_access(self) -> DiskPersistence:
+        return self._local
+
+    def construct_random_path(
+        self, persist: DataPersistence, file_path_or_file_name: typing.Optional[str] = None
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:267]: can file_path_or_file_name be `os.PathLIke`?

File: flytekit/core/data_persistence.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+        and allows for disambiguating paths within the same directory.
+
+        Use file_path_or_file_name, when you want a random directory, but want to preserve the leaf file name
+        """
+        return self.construct_random_path(self._default_remote, file_path_or_file_name)
+
+    def get_random_remote_directory(self):
+        return self.get_random_remote_path(None)
+
+    def get_random_local_path(self, file_path_or_file_name: typing.Optional[str] = None) -> str:
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:293]: can file_path_or_file_name be `os.PathLIke`?

File: flytekit/core/data_persistence.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+        return self.put_data(local_path, remote_path, is_multipart=True)
+
+    def get_data(self, remote_path: str, local_path: str, is_multipart=False):
+        """
+        :param Text remote_path:
+        :param Text local_path:
+        :param bool is_multipart:
+        """
+        try:
+            with PerformanceTimer("Copying ({} -> {})".format(remote_path, local_path)):
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:343]:
```suggestion
            with PerformanceTimer(f"Copying ({remote_path} -> {local_path})":
```

File: flytekit/core/data_persistence.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+        except Exception as ex:
+            raise FlyteAssertion(
+                "Failed to get data from {remote_path} to {local_path} (recursive={is_multipart}).\n\n"
+                "Original exception: {error_string}".format(
+                    remote_path=remote_path,
+                    local_path=local_path,
+                    is_multipart=is_multipart,
+                    error_string=str(ex),
+                )
+            )
```
Comment (cosmicBboy) [inline flytekit/core/data_persistence.py:354]:
```suggestion
            raise FlyteAssertion(
                f"Failed to get data from {remote_path} to {local_path} (recursive={is_multipart}).\n\n"
                f"Original exception: {str(ex)}"
            )
```

File: flytekit/extras/persistence/http.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super(HttpPersistence, self).__init__(name="http/https", *args, **kwargs)
+
+    def exists(self, path: str):
+        rsp = _requests.head(path)
+        allowed_codes = {
+            type(self)._HTTP_OK,
+            type(self)._HTTP_NOT_FOUND,
+            type(self)._HTTP_FORBIDDEN,
```
Comment (cosmicBboy) [inline flytekit/extras/persistence/http.py:31]: is the `type` call necessary? (see below as well)

File: flytekit/extras/persistence/http.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        if rsp.status_code not in allowed_codes:
+            raise _user_exceptions.FlyteValueException(
+                rsp.status_code,
+                "Data at {} could not be checked for existence. Expected one of: {}".format(path, allowed_codes),
```
Comment (cosmicBboy) [inline flytekit/extras/persistence/http.py:36]:
```suggestion
                f"Data at {path} could not be checked for existence. Expected one of: {allowed_codes}",
```

File: flytekit/extras/persistence/s3_awscli.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    if "ACL" in extra_args:
+        cmd += ["--acl", extra_args["ACL"]]
+    return cmd
+
+
+class S3Persistence(DataPersistence):
+    """
+    DataPersistence plugin for AWS S3 (and Minio). Use aws cli to manage the transfer. The binary needs to be installed
+    separately
+    """
```
Comment (cosmicBboy) [inline flytekit/extras/persistence/s3_awscli.py:61]:
```suggestion
    """
    DataPersistence plugin for AWS S3 (and Minio). Use aws cli to manage the transfer. The binary needs to be installed
    separately
    
    .. prompt::
       pip install awscli
    """
```

File: flytekit/extras/persistence/s3_awscli.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    def __init__(self, default_prefix: Optional[str] = None):
+        super().__init__(name="awscli-s3", default_prefix=default_prefix)
+
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the AWS cli is present
+        """
+        if not shell_which(S3Persistence._AWS_CLI):
+            raise _FlyteUserException("AWS CLI not found at Please install.")
```
Comment (cosmicBboy) [inline flytekit/extras/persistence/s3_awscli.py:76]:
```suggestion
            raise _FlyteUserException("AWS CLI not found! Please install it with `pip install awscli`.")
```

File: flytekit/extras/persistence/gcs_gsutil.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    def __init__(self, default_prefix: typing.Optional[str] = None):
+        super(GCSPersistence, self).__init__(name="gcs-gsutil", default_prefix=default_prefix)
+
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the `gsutil` cli is present
+        """
+        if not shell_which(GCSPersistence._GS_UTIL_CLI):
+            raise _FlyteUserException("gsutil (gcloud cli) not found! Please install.")
```
Comment (cosmicBboy) [inline flytekit/extras/persistence/gcs_gsutil.py:44]:
```suggestion
            raise _FlyteUserException("gsutil (gcloud cli) not found! Please install it with `pip install gsutil`.")
```

File: flytekit/extras/persistence/http.py (hunk 13 above)
Comment (kumare3) [reply flytekit/extras/persistence/http.py:31]: nope, this is legacy code, removing - good call

File: flytekit/extras/persistence/http.py (hunk 14 above)
Comment (kumare3) [reply flytekit/extras/persistence/http.py:36]: done

File: docs/source/data.extend.rst [hunk 18]
```diff
@@ -0,0 +1,14 @@
+###################
+Extend Type System
```
[the remark this answers is not in the document]
Comment (kumare3) [reply docs/source/data.extend.rst:2]: done

File: flytekit/core/data_persistence.py (hunk 1 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:13]: done

File: flytekit/core/data_persistence.py (hunk 2 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:105]: done

File: flytekit/core/data_persistence.py (hunk 3 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:50]: done

File: flytekit/core/data_persistence.py (hunk 4 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:81]: done

File: flytekit/core/data_persistence.py (hunk 7 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:240]: done

File: flytekit/core/data_persistence.py (hunk 5 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:119]: done

File: flytekit/core/data_persistence.py (hunk 6 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:164]: done

File: flytekit/core/data_persistence.py (hunk 8 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:256]: sure

File: flytekit/core/data_persistence.py (hunk 9 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:267]: not necessarily

File: flytekit/core/data_persistence.py (hunk 10 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:293]: not necessarily. Its actually a string

File: flytekit/core/data_persistence.py (hunk 11 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:343]: done

File: flytekit/core/data_persistence.py (hunk 12 above)
Comment (kumare3) [reply flytekit/core/data_persistence.py:354]: done

**Review by wild-endeavor — DISMISSED** [review]

**Comment by kumare3** [comment]
the fsspec file-system is not ready, because of a bug in fsspec - https://github.com/intake/filesystem_spec/issues/724
cc @tekumara. Not sure, if fsspec is used to copy multiple files.

**Comment by kumare3** [comment]
@tekumara, nevermind, I made a workaround, so fsspec will work in compliant mode, but for performance the bug will need to be fixed.

**Review by wild-endeavor — DISMISSED** [review]

**Comment by kumare3** [comment]
cc @wild-endeavor this is ready for review

**Comment by wild-endeavor** [comment]
Not sure how to fix this, but the error is ` ModuleNotFoundError: No module named 'flytekitplugins.fsspec'`.  I suspect what's happening is that the github workflow is running `make -C plugins install-all-dev` to install the plugins.

We could add the fsspec entry into the `plugins/setup.py` file so that `install-all-dev` picks it up, but I don't think we should.  If we do that, then developers by default when they run that command, will get this plugin, which would interfere with the base aws/gcp plugin.  ie, plugins that are loaded by default should not be installed by default.

we can add a separate command after [this](https://github.com/flyteorg/flytekit/blob/master/.github/workflows/pythonbuild.yml#L37)?

**Comment by kumare3** [comment]
@wild-endeavor This can create other problems, if we add another data plugin or the like. 
Should each plugin be an isolated package, complete with its own tests?
I know you said, we cannot do this, can you remind me why not?

File: flytekit/extras/persistence/s3_awscli.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+            file_path,
+        ]
+        try:
+            _update_cmd_config_and_execute(cmd)
+            return True
+        except Exception as ex:
+            # The s3api command returns an error if the object does not exist. The error message contains
+            # the http status code: "An error occurred (404) when calling the HeadObject operation: Not Found"
+            #  This is a best effort for returning if the object does not exist by searching
+            # for existence of (404) in the error message. This should not be needed when we get off the cli and use lib
```
Comment (tekumara) [inline flytekit/extras/persistence/s3_awscli.py:118]: is the plan to replace the cli with fsspec using s3fs? or a boto implementation?

**Review by tekumara — DISMISSED** [review]
Overall LGTM ❤️ just added some minor comments .... like @wild-endeavor I ran into some issues getting the plugins installed so I could run the tests.

File: flytekit/extras/persistence/s3_awscli.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the AWS cli is present
+        """
+        if not shell_which(S3Persistence._AWS_CLI):
+            raise _FlyteUserException("AWS CLI not found! Please install it with `pip install awscli`.")
+
+    @staticmethod
+    def _split_s3_path_to_bucket_and_key(path: str) -> (str, str):
```
Comment (tekumara) [inline flytekit/extras/persistence/s3_awscli.py:84]: To fix the syntax error in mypy/pyright:
```suggestion
    def _split_s3_path_to_bucket_and_key(path: str) -> Tuple[str, str]:
```

**Review by wild-endeavor — DISMISSED** [review]

File: flytekit/core/data_persistence.py [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+        return os.path.join(*args)
+
+
+class FileAccessProvider(object):
+    """
+    This is the class that is available through the FlyteContext and can be used for persisting data to the remote
+    durable store.
+    """
+
+    def __init__(self, local_sandbox_dir: Union[str, os.PathLike], raw_output_prefix: str):
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/core/data_persistence.py:239]: ya we call it the output prefix or local data dir

File: flytekit/extras/persistence/http.py [hunk 22]
```diff
@@ -0,0 +1,68 @@
+import os
+import pathlib
+
+import requests as _requests
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/extras/persistence/http.py:4]: haha, this is an artifact of some previous work in flytekit. the reasoning was to avoid pollution of the import space.

[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/extras/persistence/http.py:4]: in the new flytekit, we should just get rid of it, I will, good call out

File: flytekit/core/data_persistence.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/core/data_persistence.py:239]: i guess i can update it, but would like to do that as a separate PR

File: flytekit/core/data_persistence.py [hunk 23]
```diff
@@ -0,0 +1,377 @@
...
+            else:
+                # Emulate s3's flat storage by automatically creating directory path
+                self._make_local_path(os.path.dirname(self.strip_file_header(to_path)))
+                # Write the object to a local file in the sandbox
+                copyfile(self.strip_file_header(from_path), self.strip_file_header(to_path))
+
+    def construct_path(self, _: bool, add_prefix: bool, *args) -> str:
+        # Ignore add_protocol for now. Only complicates things
+        if add_prefix:
+            return os.path.join(self.default_prefix, *args)
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/core/data_persistence.py:229]: done

File: flytekit/core/data_persistence.py [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+
+    @abstractmethod
+    def put(self, from_path: str, to_path: str, recursive: bool = False):
+        """
+        Stores data from from_path and writes to the given to_path (from_path is locally accessible)
+        """
+        pass
+
+    @abstractmethod
+    def construct_path(self, add_protocol: bool, add_prefix: bool, *paths) -> str:
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/core/data_persistence.py:95]: done

File: flytekit/extras/persistence/s3_awscli.py [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+        S3Persistence._check_binary()
+        cmd = [S3Persistence._AWS_CLI, "s3", "cp"]
+        if recursive:
+            cmd += ["--recursive"]
+        cmd.extend(_extra_args(extra_args))
+        cmd += [from_path, to_path]
+        return _update_cmd_config_and_execute(cmd)
+
+    def construct_path(self, add_protocol: bool, add_prefix: bool, *paths) -> str:
+        paths = list(paths)  # make type check happy
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply flytekit/extras/persistence/s3_awscli.py:153]: unfortunately we're not really strictly following mypy right now... for reason I forget (?) @kumare3 @wild-endeavor

[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/extras/persistence/s3_awscli.py:153]: we should - we just haven't time to go through the whole codebase and clean everything up.

**Review by wild-endeavor — APPROVED** [review]

File: flytekit/extras/persistence/http.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/extras/persistence/http.py:4]: done for all new files

File: plugins/flytekit-data-fsspec/setup.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+plugin_requires = ["flytekit>=0.21.3,<1.0.0", "fsspec"]
+
+__version__ = "0.0.0+develop"
+
+setup(
+    name=microlib_name,
+    version=__version__,
+    author="flyteorg",
+    author_email="admin@flyte.org",
+    description="This package holds Hive plugins for flytekit",
```
[the remark this answers is not in the document]
Comment (kumare3) [reply plugins/flytekit-data-fsspec/setup.py:16]: done

File: flytekit/extras/persistence/s3_awscli.py (hunk 19 above)
Comment (kumare3) [reply flytekit/extras/persistence/s3_awscli.py:118]: we cannot replace, as there are existing users of this. but users can optionally install `pip install flytekitplugins-data-fsspec` or other such solutions in the future

File: flytekit/extras/persistence/s3_awscli.py (hunk 20 above)
Comment (kumare3) [reply flytekit/extras/persistence/s3_awscli.py:84]: Done

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — DISMISSED** [review]

**Review by tekumara — APPROVED** [review]

**Comment by samhita-alla** [comment]
@kumare3 / @eapolinario Tests are passing now! Modified the Great Expectations plugin to accept remote FlyteFiles only. If the user wants to give a local dataset file, it can only be given as a string. I've also removed the concerned local FlyteFile tests. See this commit: https://github.com/flyteorg/flytekit/pull/559/commits/35920c730dfe135d7e6e9b9786c5dcb1fabe7ca5

When given a local dataset file that has a relative path, FlyteFile is not working because of this:
```python
    def is_remote(path: Union[str, os.PathLike]) -> bool:
        """
        Deprecated. Lets find a replacement
        """
        return not (path.startswith("/") or path.startswith("file://"))
```

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
